File: rustt/src/array/two_zero_nine.rs
use std::result;

/*
 * @lc app=leetcode.cn id=209 lang=rust
 *
 * [209] 长度最小的子数组
 */
pub struct Solution;

// @lc code=start
impl Solution {
    pub fn min_sub_array_len(target: i32, nums: Vec<i32>) -> i32 {
        let (mut res, mut sub_length) = (i32::MAX, 0);
        let (mut i, mut sum) = (0, 0);
        for (index, value) in nums.iter().enumerate() {
            sum += value;
            while sum >= target {
                //* get the sub length */
                sub_length = (index - i + 1) as i32;
                //* compare res length and sub length */
                res = res.min(sub_length as i32);
                sum -= nums[i];
                i += 1;
            }
        }
        if res == i32::MAX {
            return 0;
        }
        res as i32
    }
}
// @lc code=end

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_min_sub_array_len() {
        let nums = vec![2, 3, 1, 2, 4, 3];
        let target = 100;
        let res = Solution::min_sub_array_len(target, nums);
        assert_eq!(res, 0);
    }
}

File: c/src/Array/two_zero_nine.c
/*
 * @lc app=leetcode.cn id=209 lang=c
 *
 * [209] 长度最小的子数组
 */

// @lc code=start
int minSubArrayLen(int target, int *nums, int numsSize) {
    int result = __INT_MAX__;
    int left = 0;
    int sum = 0;
    for (int right = 0; right < numsSize; right++) {
        sum += nums[right];
        while (sum >= target) {
            int len = right - left + 1;
            result = len <= result ? len : result;
            sum -= nums[left++];
        }
    }
    return result == __INT_MAX__ ? 0 : result;
}
// @lc code=end
